File: go/src/prime_factors/prime_factors.go
package prime_factors

func Factors(num int) []int {
    return factors(num, 2)
}

func factors(num int, candidate int) []int {
    switch {
        case num == 1: return []int{}
        case num % candidate == 0: return append([]int{candidate}, factors(num / candidate, candidate)...)
        default: return factors(num, candidate + 1)
    }
}

File: scala/src/main/scala/PrimeFactors.scala
package com.crebma.primeFactors

object PrimeFactors {

  /**def factors(num: Int) : List[Int] = {
    var number = num
    var primes = List[Int]()
    var candidate = 2
    
    while (number > 1) {
        while (number % candidate == 0) {
            primes = primes :+ candidate
            number = number / candidate
        }
        
        candidate = candidate + 1
    }
    
    primes
  }*/
  
  def factors(num: Int, candidate: Int = 2) : List[Int] = num match  {
    case 1 => List[Int]()
    case x if x % candidate == 0 => candidate :: factors(x / candidate, candidate)
    case x => factors(x, candidate + 1)
  }
}
